### PR TITLE
Allow autocomplete based on phone sync

### PR DESCRIPTION
**Pull request by nickvergessen**
- [x] Based on #26029 
- [x] Store phonebook sync matches in a database
- [x] Remove "known user" matches when the user changes their phone number
- [x] Remove any matches when the user is deleted
- [x] Add a config option to "Restrict username autocompletion to users with phone sync matches" in the sharing settings
    - [x] (needs some thinking how we can phrase it so it works with the existing options in a good way)
- [x] Respect that setting on the autocomplete endpoint.
- [x] Write integration tests to assure this doesn't break

### CODE REVIEW

**Review by ChristophWurst — APPROVED** [review]
I squinted but the code looks fine

File: lib/private/KnownUser/KnownUserMapper.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+	/**
+	 * @param IDBConnection $db
+	 */
+	public function __construct(IDBConnection $db) {
+		parent::__construct($db, 'known_users', KnownUser::class);
+	}
+
+	/**
+	 * @param string $knownTo
```
Comment (PVince81) [inline lib/private/KnownUser/KnownUserMapper.php:42]: add comment that this is the user id (and not contact id or whatever else) ? (also in various other places)
or rename to `$knownToUserId` ?

and maybe add a description to make it easier to distinguish between "knownTo" and "knownUser".

File: lib/private/KnownUser/KnownUserMapper.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	public function deleteKnownUser(string $knownUser): int {
+		$query = $this->db->getQueryBuilder();
+		$query->delete($this->getTableName())
+			->where($query->expr()->eq('known_user', $query->createNamedParameter($knownUser)));
+
+		return (int) $query->execute();
+	}
+
+	/**
+	 * @param string $knownTo
```
Comment (PVince81) [inline lib/private/KnownUser/KnownUserMapper.php:66]: Returns all user ids that... the brain gymnastics hurts a bit, a description would be good here

File: lib/private/KnownUser/KnownUserService.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OC\KnownUser;
+
+class KnownUserService {
```
Comment (PVince81) [inline lib/private/KnownUser/KnownUserService.php:26]: great idea to introduce this :+1:

File: lib/private/KnownUser/KnownUserService.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+	public function deleteKnownTo(string $knownTo): int {
+		return $this->mapper->deleteKnownTo($knownTo);
+	}
+
+	public function deleteKnownUser(string $knownUser): int {
+		return $this->mapper->deleteKnownUser($knownUser);
+	}
+
+	public function storeIsKnownToUser(string $knownTo, string $knownUser): void {
```
Comment (PVince81) [inline lib/private/KnownUser/KnownUserService.php:44]: again confused who is the user who is known to who.
as an API consumer I wouldn't know where to pass which user and would resort to trial and error

**Review by PVince81 — APPROVED** [review]
:+1: code looks fine with minor comments

and one thing that disturbed me most was the naming of $knownTo and $knownUser which are easy to mix up

File: lib/private/Contacts/ContactsMenu/ContactsStore.php [hunk 5]
```diff
@@ -165,24 +165,16 @@ private function filterContacts(IUser $self,
...
 					}
 
-					$filterUser = true;
-
 					$mailAddresses = $entry->getEMailAddresses();
 					foreach ($mailAddresses as $mailAddress) {
 						if ($mailAddress === $filter) {
-							$filterUser = false;
-							break;
+							return true;
```
Comment (PVince81) [inline lib/private/Contacts/ContactsMenu/ContactsStore.php:171]: it feels wrong now to return directly here. if we said we don't want to filter, it seems the old code flow would continue into later if statements.

sorry if my former misunderstanding and review comment lead to this change :-/

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by skjnldsv** [comment]
@nickvergessen Documentation missing?

**Comment by nickvergessen** [comment]
How to upload your phone's addressbook? Or the admin setting?

**Comment by skjnldsv** [comment]
> How to upload your phone's addressbook? Or the admin setting?

admin setting
https://github.com/nextcloud/server/blob/d9f85220038b2421870a36f9b6a860d6444a988a/apps/settings/lib/Settings/Admin/Sharing.php#L88-L94

**Comment by nickvergessen** [comment]
Added the new checkbox in https://github.com/nextcloud/documentation/pull/11241 and updated the screenshot and labels that were brought in by other changes in the meantime

**Pull request merged**